File: SegmentTree.java
import java.util.*;


class SegmentTree {
    private static int sz, hei;

    private static boolean balance;

    private static ArrayList<Seg_Node> t;
    public SegmentTree(int N, int[] val){
//        sz = N;
        int S = 3*N;

        // Comon java, why do you need me to manually create this....get ur sh*t together
        for(int i = 0; i<S; ++i){
            t.add(new Seg_Node());
        }
    }



    private static void psUp(int idx){

        if(t.get(idx).getL()!=t.get(idx).getR()){
            t.get(idx).setGCD(Utilities.GCD(t.get(idx<<1).getGCD(),t.get(idx<<1|1).getGCD()));

            t.get(idx).setH(Math.max(t.get(idx<<1).getH(),t.get(idx<<1|1).getH()));

            t.get(idx).setMX(Math.max(t.get(idx<<1).getMX(),t.get(idx<<1|1).getMX()));

            t.get(idx).setMN(Math.max(t.get(idx<<1).getMN(),t.get(idx<<1|1).getMN()));

            t.get(idx).setH(Math.max(t.get(idx<<1).getH(),t.get(idx<<1|1).getH()));

            t.get(idx).setSz(t.get(idx<<1).getSz()+t.get(idx<<1|1).getSz()+1);

            t.get(idx).setBal(t.get(idx<<1).getBal()&& t.get(idx<<1|1).getBal()&&(Math.abs(t.get(idx<<1).getSz()-t.get(idx<<1|1).getSz())<=1));
        }

    }

    static void bld(int[] val, int idx, int l, int r){
        t.get(idx).setL(l);
        t.get(idx).setR(r);
        if(l==r){
//            t.get(idx).setVal(val[l]);
            t.set(idx, new Seg_Node(l,r,val[l],val[l],val[l],1,true));
            return;
        }

        int mid = l + (r - l)>>1;
        bld(val,idx<<1,l,mid);
        bld(val,idx<<1|1,mid+1,r);

        psUp(idx);
    }

    static void updI(int idx, int targ, int val){
        t.set(idx,new Seg_Node(t.get(idx).getL(),t.get(idx).getR(),0,0,0,t.get(idx).getH(),t.get(idx).getBal()));
        if(t.get(idx).getL()==t.get(idx).getR()&&t.get(idx).getL()==targ){
            t.get(idx).setGCD(val);
            t.get(idx).setMN(val);
            t.get(idx).setMX(val);
        }else{
            int mid = t.get(idx).getL() + (t.get(idx).getR() - t.get(idx).getL())>>1;
            if(targ<=mid)updI(idx<<1,targ,val);
            else updI(idx<<1|1,targ,val);
            psUp(idx);
        }
    }

    static int getSz() {
        return t.get(1).getSz();
    }

    static boolean getBal(){
        return balance;
    }

    static int getHei(){
        return hei;
    }

    
}


class BST{

}

File: SegmentTree.cpp
//
// Created by michael on 02/05/19.
//

#include "SegmentTree.h"


SegNode SegmentTree ::getidx(int idx) {
    return tree[idx];
}

int SegmentTree ::get(int idx, int l, int r, int (*op)(int, int), char cmd) {
    if(tree[idx].getL() == l && tree[idx].getR() == r){
        if(cmd == 'n'){
            return tree[idx].getMN();
        }
        if(cmd == 'x'){
            return tree[idx].getMX();
        }
        if(cmd == 'g'){
            return tree[idx].getGcd();
        }
    }

    int m = mid(tree[idx].getL(),tree[idx].getR());

    if(r <= m){
        return get(l(idx), l, r,op,cmd);
    }

    if(l > m){
        return get(r(idx), l, r, op, cmd);
    }

    return op(get(l(idx), l, m, op, cmd), get(r(idx), m+1, r, op, cmd));
}



int SegmentTree :: getMN(int l, int r){
    int (*minF)(int,int) = min;
    return get(1, l, r, minF, 'n');
}

int SegmentTree ::getMX(int l, int r) {
    int (*maxF)(int,int) = max;
    return get(1, l, r, maxF, 'x');
}

int SegmentTree ::getGCD(int l, int r) {
    int(*gcdF)(int, int) = gcd;
    return get(1, l, r, gcdF, 'g');
}

void SegmentTree ::psUp(int idx) {
    if(tree[idx].getL() != tree[idx].getR()){
        tree[idx].setGCD(gcd(tree[l(idx)].getGcd(), tree[r(idx)].getGcd()));

        tree[idx].setMX(max(tree[l(idx)].getMX(), tree[r(idx)].getMX()));

        tree[idx].setMN(min(tree[l(idx)].getMN(), tree[r(idx)].getMN()));
    }

}

void SegmentTree ::bld(int idx, int l, int r) {
    tree[idx].setL(l);
    tree[idx].setR(r);

    if(l == r){
        int val;
        scanf("%d",&val);
        tree[idx].setMN(val);
        tree[idx].setMX(val);
        tree[idx].setGCD(val);
        return;
    }

    int m = mid(l, r);

    bld(l(idx), l, m);
    bld(r(idx), m+1, r);

    psUp(idx);
}

void SegmentTree ::upd(int idx, int l, int r, int val) {
    if(tree[idx].getL() == l && tree[idx].getR() == r){
        tree[idx].setMN(val);
        tree[idx].setMX(val);
        tree[idx].setGCD(val);
        return;
    }

    int m = mid(l,r);

    if(tree[idx].getL() > r){
        upd(l(idx), l, m, val);
    }else{
        upd(r(idx), m+1, r, val);
    }

    psUp(idx);
}

void SegmentTree :: upd(int idx, int val){
    upd(idx, 1, N, val);
}

void SegmentTree ::print(int idx, int l, int r) {
    if(tree[idx].getL() == tree[idx].getR()){
        printf("Node %d, left range %d, right range %d\n", idx, tree[idx].getL(),tree[idx].getR());
        printf("MX %d, MN %d, GCD %d\n",tree[idx].getMX(), tree[idx].getMN(), tree[idx].getGcd());
        return;
    }

    print(l(idx), l, r);
    print(r(idx), l, r);

    printf("Node %d, left range %d, right range %d\n", idx, tree[idx].getL(),tree[idx].getR());
    printf("MX %d, MN %d, GCD %d\n",tree[idx].getMX(), tree[idx].getMN(), tree[idx].getGcd());

}

void SegmentTree :: print(){
    print(1,1,N);
}

